File: linkedlist/cpp/linkedlist.cpp
#include "linkedlist.hpp"
#include <iostream>
#include <stack>

Node::Node() {
    this->prev = nullptr;
    this->next = nullptr;
    this->value = 0;
}

Node::Node(int value) {
    this->prev = nullptr;
    this->next = nullptr;
    this->value = value;
}

int Node::getValue() { return this->value; }
Node *Node::getNext() { return this->next; }
Node *Node::getPrev() { return this->prev; }

void Node::setValue(int value) { this->value = value; }
void Node::setNext(Node *next) { this->next = next; }
void Node::setPrev(Node *prev) { this->prev = prev; }

LinkedList::LinkedList() {
    this->head = nullptr;
    this->tail = nullptr;
    this->len = 0;
}

LinkedList::LinkedList(Node *head, Node *tail) {
    this->head = head;
    this->tail = tail;
    Node *temp = head;
    this->len = 0;
    while (temp != tail) {
        temp = temp->getNext();
        this->len++;
    }
}

Node *LinkedList::getHead() { return this->getHead(); }
Node *LinkedList::getTail() { return this->getTail(); }

unsigned int LinkedList::length() { return this->len; }
int LinkedList::begin() { return this->getHead()->getValue(); }
int LinkedList::end() { return this->getTail()->getValue(); }

void LinkedList::insert(int index, int value) {
    if (index < 0 && index > this->length()) {
        std::cerr << "Index out of range!" << std::endl;
        return;
    } else if (index == 0) {
        this->push_front(value);
    } else if (index == this->length()) {
        this->push_back(value);
    } else {
        int curr = 0;
        Node *temp = this->head;
        while (index != curr) {
            temp = temp->getNext();
            curr++;
        }
        Node *newNode = new Node(value);
        newNode->setPrev(temp->getPrev());
        newNode->setNext(temp->getNext());
        temp->getPrev()->setNext(newNode);
        temp->getNext()->setPrev(newNode);
    }
    this->len++;
}

void LinkedList::push_back(int value) {
    Node *newNode = new Node(value);
    if (this->length() == 0) {
        this->head = this->tail = newNode;
    }
    this->tail->setNext(newNode);
    newNode->setPrev(this->tail);
    this->tail = newNode;
    this->len++;
}

int LinkedList::pop_back() {
    if (this->length() == 0) {
        std::cerr << "List is empty!\n";
        return 0;
    } else if (this->length() == 1) {
        delete this->head;
        this->head = nullptr;
        this->tail = nullptr;
    } else {
        Node *temp = this->tail;
        int value = this->tail->getValue();
        this->tail = this->tail->getPrev();
        this->tail->setNext(nullptr);
        delete temp;
        return value;
    }
    this->len--;
    return 0;
}

void LinkedList::push_front(int value) {
    Node *newNode = new Node(value);
    if (this->head == nullptr && this->tail == nullptr) {
        this->head = this->tail = newNode;
    }
    newNode->setNext(this->head);
    this->head->setPrev(newNode);
    this->head = newNode;
    this->len++;
}

int LinkedList::pop_front() {
    if (this->length() == 0) {
        std::cerr << "List is empty!\n";
        return 0;
    } else if (this->length() == 1) {
        delete this->head;
        this->head = nullptr;
        this->tail = nullptr;
    } else {
        Node *temp = this->head;
        int value = this->head->getValue();
        this->head = this->head->getNext();
        this->head->setPrev(nullptr);
        delete temp;
        return value;
    }
    this->len--;
    return 0;
}

int LinkedList::at(unsigned int index) {
    if (index >= this->length()) {
        std::cerr << "Index out of range!" << std::endl;
        return 0;
    }
    int curr = 0;
    Node *temp = this->head;
    while (curr != index) {
        temp = temp->getNext();
        curr++;
    }
    return temp->getValue();
}

int LinkedList::deleteNode(unsigned int index) {
    if (index >= this->length()) {
        std::cerr << "Index out of range!" << std::endl;
        return 0;
    }
    int curr = 0, value = 0;
    Node *temp = this->head;
    while (curr != index) {
        temp = temp->getNext();
        curr++;
    }
    temp->getPrev()->setNext(temp->getNext());
    temp->getNext()->setPrev(temp->getPrev());
    value = temp->getValue();
    delete temp;
    this->len--;
    return value;
}

int LinkedList::find(int target) {
    if (this->length() == 0) {
        std::cerr << "List is empty!" << std::endl;
        return -1;
    }
    Node *temp = this->head;
    int curr = 0;
    while (temp != this->tail) {
        if (temp->getValue() == target) {
            return curr;
        }
        temp = temp->getNext();
        curr++;
    }
    return -1;
}

void LinkedList::print() {
    std::cout << "[";
    Node *temp = this->head;
    while (temp != this->tail) {
        std::cout << temp->getValue() << ", ";
        temp = temp->getNext();
    }
    std::cout << temp->getValue() << "]" << std::endl;
}

void LinkedList::reverse() {
    Node *temp = this->head;
    std::stack<typeof(this->head->getValue())> stack;

    while (temp != nullptr) {
        stack.push(temp->getValue());
        temp = temp->getNext();
    }

    temp = this->head;
    while (temp != nullptr) {
        temp->setValue(stack.top());
        stack.pop();
        temp = temp->getNext();
    }
}

File: linkedlist/py/lib/linkedlist.py
from typing import Optional
from lib.node import Node


class LinkedList:
    def __init__(self, head: Optional[Node] = None,
                 tail: Optional[Node] = None):
        """
        params:
            self: LinkedList
            head: Node | None
            tail: Node | None
        """
        self.head: Optional[Node] = head
        self.tail: Optional[Node] = tail
        self.length = 0

        # checks if length is 0 or not
        if head is not None:
            self.length = 1
            temp = head
            # iterate until temp becomes same as tail or None
            while temp is not tail and temp is not None:
                temp = temp.next
                self.length += 1

    # special method to overload [] to set value at index [x]
    def __setitem__(self, key: int, item: int) -> bool:
        """
        params:
            self: LinkedList
            index: int
            value: int
        return: bool
        """
        if key >= self.length:
            raise IndexError("Index out of bound!\n")

        curr: int = 0
        temp = self.head

        # iterate until curr index is equal to index or temp becomes None
        while curr is not key and temp is not None:
            temp = temp.next
            curr += 1

        # update value at the curr index and return true on success
        if curr is key and temp is not None:
            temp.value = item
            return True

        return False

    # special method to overload [] to get value at index [x]
    def __getitem__(self, index: int) -> Optional[int]:
        """
        params:
            self: LinkedList
            index: int
        return: int | None
        """
        current: int = 0
        temp = self.head

        # iterate until curr index is equal to index or temp becomes None
        while current is not index and temp is not None:
            temp = temp.next
            current += 1
            # checks if current index is same as required index
            if current is index and temp is not None:
                return temp.value
        return None

    # special method to overload [] to delete value at index [x]
    def __delitem__(self, key: int) -> bool:
        """
        params:
            self: LinkedList
            index: int
        return: int | None
        """
        if key >= self.length:
            raise IndexError("Index out of bound!\n")

        if key == 0 and self.head is not None:
            temp = self.head
            self.head = self.head.next
            del temp
            return True

        if key == self.length-1 and self.tail is not None:
            temp = self.tail
            self.tail = self.tail.prev
            del temp
            return True

        currItem: int = 0
        temp = self.head

        # iterate until curr index is equal to index or temp becomes None
        while currItem is not key and temp is not None:
            temp = temp.next
            currItem += 1

        # update value at the curr index and return true on success
        if temp is not None and temp.prev is not None and temp.next is not None:
            temp.prev.next = temp.next
            temp.next.prev = temp.prev
            del temp
            return True
        return False

    def __len__(self):
        return self.length

    def __str__(self):
        if self.head is None:
            return '<LinkedList>[]'

        str = '<LinkedList>['
        temp = self.head
        while temp.next is not None:
            str += f"{temp.value} <=> "
            temp = temp.next
        str += f"{temp.value}"
        str += ']'
        return str

    def insert(self, index: int, value: int) -> bool:
        """
        params:
            index: int
            value: int
        return: bool
        """
        if index > self.length:
            return False

        if index == 0:
            return self.push_head(value)

        if index == self.length:
            return self.push_tail(value)

        curr: int = 0
        temp = self.head
        while curr is not index and temp is not None:
            temp = temp.next
            curr += 1

        if curr is index and temp is not None and temp.prev is not None:
            newNode = Node(value)
            newNode.next = temp
            newNode.prev = temp.prev
            temp.prev.next = newNode
            temp.prev = newNode

        return False

    def delete_node(self, index: int) -> Optional[int]:
        """
        params:
            self: LinkedList
            index: int
        return: bool
        """
        if index >= self.length:
            return None

        if index == 0:
            return self.pop_head()

        if index == self.length-1:
            return self.pop_tail()

        curr: int = 0
        temp = self.head
        while curr is not index and temp is not None:
            temp = temp.next
            curr += 1

        if temp is not None and temp.prev is not None and temp.next is not None:
            temp.prev.next = temp.next
            temp.next.prev = temp.prev
            value = temp.value
            del temp
            return value

        return False

    def get_head(self) -> Optional[int]:
        """
        params:
            self: LinkedList
        return: int | None
        """
        if self.head is not None:
            return self.head.value
        return None

    def push_head(self, item: int) -> bool:
        """
        params:
            self: LinkedList
            item: int
        return: bool
        """

        newNode = Node(item)

        if self.head is not None:
            newNode.next = self.head
            self.head.prev = newNode
            self.head = newNode
        else:
            self.head = self.tail = newNode

        return True

    def pop_head(self) -> Optional[int]:
        """
        params:
            self: LinkedList
        return: int | None
        """
        if self.head is not None:
            value = self.head.value
            self.head = self.head.next
            if self.head is not None:
                del self.head.prev
                self.head.prev = None
            return value
        return None

    def get_tail(self) -> Optional[int]:
        """
        params:
            self:LinkedList
        return: int | None
        """
        if self.tail is not None:
            return self.tail.value
        return None

    def push_tail(self, item: int) -> bool:
        """
        params:
            self: LinkedList
            item: int
        return: bool
        """
        newNode = Node(item)

        if self.tail is not None:
            newNode.prev = self.tail
            self.tail.next = newNode
            self.tail = newNode
        else:
            self.head = self.tail = newNode

        return True

    def pop_tail(self) -> Optional[int]:
        """
        params:
            self: LinkedList
        return: int | None
        """
        if self.tail is not None:
            value = self.tail.value
            self.tail = self.tail.prev
            if self.tail is not None:
                del self.tail.next
                self.tail.next = None
            return value
        return None
